schema/validator.go

```diff
@@ -54,6 +55,28 @@ func (v Validator) Validate(src io.Reader) error {
 			"schema %s: unable to validate", v)
 	}
 
+	if v == MediaTypeManifest {
```

Comment:
On Thu, Sep 08, 2016 at 12:29:29AM -0700, xiekeyang wrote:
> > +	if v == MediaTypeManifest {
> 
> @wking 
> > ?  ‘oci-image-tool validate …’ calls both image.Validate(Layout) [1]
> > and schema.MediaType(Manifest|ManifestList|Config).Validate [2].  I
> > expect we should be calling the schema validation from the image
> > validation functions, but either way, code in the image validation
> > functions is getting called by ‘oci-image-tool validate …’.
> 
> It is some hard to do based on current code.
> What you mentioned [1] is on package order main -> image -> schema;
> [2] is main -> schema;
> I think it had better to add code to schema package, and better to
> only one place to be called by up layer.

I'd rather not tie the JSON Schema too tightly here.  Having the
schemas and schema-only tooling in their own layer makes it easier to
share them (and there is a lot of third-party JSON Schema tooling that
can consume them, opencontainers/validator.opencontainers.org#2).

On the other hand, JSON Schema cannot represent all of our
constraints.  So I'd like another layer (currently in image/*.go) to
do the full validation.  That full-validation layer can save itself a
lot of work by leaning on JSON Schema for part of its validation, but
whether it uses JSON Schema or not is really an internal detail.

So I think we want to use only main → image → schema, but would also
be fine with main → image/validate → schema (if you want to separate
the validation Go code from the unpacking, etc., Go code).  I don't
think we want to push a lot of Go logic into the JSON Schema
directory.